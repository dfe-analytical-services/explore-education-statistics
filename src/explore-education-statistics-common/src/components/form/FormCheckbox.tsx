import classNames from 'classnames';
import React, { ChangeEvent, FocusEventHandler, memo, ReactNode } from 'react';

export type OtherCheckboxChangeProps = Pick<FormCheckboxProps, 'label'>;

export type CheckboxChangeEventHandler = (
  event: ChangeEvent<HTMLInputElement>,
  otherCheckboxProps: OtherCheckboxChangeProps,
) => void;

export interface FormCheckboxProps {
  checked?: boolean;
  defaultChecked?: boolean;
  className?: string;
  conditional?: ReactNode;
  id: string;
  hint?: string;
  label: string;
  boldLabel?: boolean;
  name: string;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  onChange?: CheckboxChangeEventHandler;
  value: string;
  disabled?: boolean;
}

const FormCheckbox = ({
  checked,
  defaultChecked,
  className,
  conditional,
  id,
  hint,
  label,
  boldLabel = false,
  name,
  onBlur,
  onChange,
  value,
  disabled = false,
}: FormCheckboxProps) => {
  return (
    <>
      <div className={classNames('govuk-checkboxes__item', className)}>
        <input
          aria-describedby={hint ? `${id}-item-hint` : undefined}
          className="govuk-checkboxes__input"
          checked={checked}
          defaultChecked={defaultChecked}
          id={id}
          name={name}
          onBlur={onBlur}
          onChange={event => {
            if (onChange) {
              onChange(event, { label });
            }
          }}
          type="checkbox"
          value={value}
          disabled={disabled}
        />
        <label className="govuk-label govuk-checkboxes__label" htmlFor={id}>
          {boldLabel ? <strong>{label}</strong> : label}
        </label>
        {hint && (
          <span
            id={`${id}-item-hint`}
            className="govuk-hint govuk-checkboxes__hint"
          >
            {hint}
          </span>
        )}
      </div>
      {conditional && (
        <div
          className={classNames('govuk-checkboxes__conditional', {
            'govuk-checkboxes__conditional--hidden': !checked,
          })}
        >
          {conditional}
        </div>
      )}
    </>
  );
};

export default memo(FormCheckbox);
