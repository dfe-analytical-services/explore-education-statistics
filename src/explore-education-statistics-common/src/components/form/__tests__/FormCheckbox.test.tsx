import { render } from '@testing-library/react';
import React from 'react';
import FormCheckbox from '../FormCheckbox';

describe('FormCheckbox', () => {
  test('renders correctly with required props', () => {
    const { container } = render(
      <FormCheckbox
        name="test"
        id="test-checkbox"
        label="Test checkbox"
        value="true"
      />,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly when `hint` is provided', () => {
    const { container, getByLabelText, getByText } = render(
      <FormCheckbox
        name="test"
        hint="Click me to proceed"
        id="test-checkbox"
        label="Test checkbox"
        value="true"
      />,
    );

    expect(getByText('Click me to proceed')).toHaveAttribute(
      'id',
      'test-checkbox-item-hint',
    );
    expect(getByLabelText('Test checkbox')).toHaveAttribute(
      'aria-describedby',
      'test-checkbox-item-hint',
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders conditional content when checked', () => {
    const { container, getByText } = render(
      <FormCheckbox
        name="test"
        id="test-checkbox"
        label="Test checkbox"
        value="true"
        checked
        onChange={() => {}}
        conditional={<p>The conditional content</p>}
      />,
    );

    expect(getByText('The conditional content').parentElement).not.toHaveClass(
      'govuk-checkboxes__conditional--hidden',
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('does not render conditional content when not checked', () => {
    const { container, getByText } = render(
      <FormCheckbox
        name="test"
        id="test-checkbox"
        label="Test checkbox"
        value="true"
        conditional={<p>The conditional content</p>}
      />,
    );

    expect(getByText('The conditional content').parentElement).toHaveClass(
      'govuk-checkboxes__conditional--hidden',
    );
    expect(container.innerHTML).toMatchSnapshot();
  });
});
