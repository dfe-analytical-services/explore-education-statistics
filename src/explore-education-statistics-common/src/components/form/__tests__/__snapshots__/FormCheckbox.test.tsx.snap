// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckbox does not render conditional content when not checked 1`] = `
<div class="govuk-checkboxes__item"
     data-testid="Checkbox item for Test checkbox"
>
  <input class="govuk-checkboxes__input"
         id="test-checkbox"
         name="test"
         type="checkbox"
         value="true"
         aria-controls="test-checkbox-conditional"
  >
  <label class="govuk-label govuk-checkboxes__label"
         for="test-checkbox"
  >
    Test checkbox
  </label>
</div>
<div class="govuk-checkboxes__conditional govuk-checkboxes__conditional--hidden"
     id="test-checkbox-conditional"
>
  <p>
    The conditional content
  </p>
</div>
`;

exports[`FormCheckbox renders conditional content when checked 1`] = `
<div class="govuk-checkboxes__item"
     data-testid="Checkbox item for Test checkbox"
>
  <input class="govuk-checkboxes__input"
         id="test-checkbox"
         name="test"
         type="checkbox"
         value="true"
         checked
         aria-controls="test-checkbox-conditional"
         aria-expanded="true"
  >
  <label class="govuk-label govuk-checkboxes__label"
         for="test-checkbox"
  >
    Test checkbox
  </label>
</div>
<div class="govuk-checkboxes__conditional"
     id="test-checkbox-conditional"
>
  <p>
    The conditional content
  </p>
</div>
`;

exports[`FormCheckbox renders correctly when \`hint\` is provided 1`] = `
<div class="govuk-checkboxes__item"
     data-testid="Checkbox item for Test checkbox"
>
  <input aria-describedby="test-checkbox-item-hint"
         class="govuk-checkboxes__input"
         id="test-checkbox"
         name="test"
         type="checkbox"
         value="true"
  >
  <label class="govuk-label govuk-checkboxes__label"
         for="test-checkbox"
  >
    Test checkbox
  </label>
  <div id="test-checkbox-item-hint"
       class="govuk-hint govuk-checkboxes__hint"
  >
    Click me to proceed
  </div>
</div>
`;

exports[`FormCheckbox renders correctly with required props 1`] = `
<div class="govuk-checkboxes__item"
     data-testid="Checkbox item for Test checkbox"
>
  <input class="govuk-checkboxes__input"
         id="test-checkbox"
         name="test"
         type="checkbox"
         value="true"
  >
  <label class="govuk-label govuk-checkboxes__label"
         for="test-checkbox"
  >
    Test checkbox
  </label>
</div>
`;
